airflow/jobs/scheduler_job.py

```diff
@@ -1705,62 +1305,216 @@ def _run_scheduler_loop(self) -> None:
-        # DAG runs in those states
-        # If a task instance is up for retry but the corresponding DAG run
-        # isn't running, mark the task instance as FAILED so we don't try
-        # to re-run it.
-        self._change_state_for_tis_without_dagrun(
-            simple_dag_bag=simple_dag_bag,
-            old_states=[State.UP_FOR_RETRY],
-            new_state=State.FAILED
-        )
-        # If a task instance is scheduled or queued or up for reschedule,
-        # but the corresponding DAG run isn't running, set the state to
-        # NONE so we don't try to re-run it.
-        self._change_state_for_tis_without_dagrun(
-            simple_dag_bag=simple_dag_bag,
-            old_states=[State.QUEUED,
-                        State.SCHEDULED,
-                        State.UP_FOR_RESCHEDULE,
-                        State.SENSING],
-            new_state=State.NONE
+                if self.executor.slots_available <= 0:
+                    # We know we can't do anything here, so don't even try!
+                    self.log.debug("Executor full, skipping critical section")
+                    return 0
+
+                timer = Stats.timer('scheduler.critical_section_duration')
+                timer.start()
+
+                # Find anything TIs in state SCHEDULED, try to QUEUE it (send it to the executor)
+                num_queued_tis = self._critical_section_execute_task_instances(session=session)
+
+                # Make sure we only sent this metric if we obtained the lock, otherwise we'll skew the
+                # metric, way down
+                timer.stop(send=True)
+            except OperationalError as e:
+                timer.stop(send=False)
+
+                # DB specific error codes:
+                # Postgres: 55P03
+                # MySQL: 3572, 'Statement aborted because lock(s) could not be acquired immediately and NOWAIT
+                #               is set.'
+                # MySQL: 1205, 'Lock wait timeout exceeded; try restarting transaction
+                #              (when NOWAIT isn't available)
+                db_err_code = getattr(e.orig, 'pgcode', None) or e.orig.args[0]
+
+                # We could test if e.orig is an instance of
+                # psycopg2.errors.LockNotAvailable/_mysql_exceptions.OperationalError, but that involves
+                # importing it. This doesn't
+                if db_err_code in ('55P03', 1205, 3572):
+                    self.log.debug("Critical section lock held by another Scheduler")
+                    Stats.incr('scheduler.critical_section_busy')
+                    return 0
+                raise
+
+            return num_queued_tis
+        finally:
+            event.remove(session.bind, 'commit', validate_commit)
+
+    def _create_dag_run(self, dag_model: DagModel, dag: DAG, session: Session) -> None:
+        """
+        Unconditionally create a DAG run for the given DAG, and update the dag_model's fields to control
```

Comment:
Unconditional here as the checks are done elsewhere/at the caller.

I'll double check the case you mention - I think it's already covered in the unit tests too

Reply:
Got it. I guess then we need some more checks in the caller--currently I haven't noticed other checks except checking the existence of `next_dagrun_create_after` value before we come here create the DAG run. The test `test_runs_respected_after_clear` I think covers part of it, it checks the scheduling part but even if it passed (which I don't think it will given that[ it returns directly when active run > max run](https://github.com/apache/airflow/pull/10956/files/f3c6cf1dc10e509da1b3b1db2a595a58abfab847#diff-c35269bcfbbe386e269ffa7487e86192R1492-R1498) rather than before[ active run += 1 after the run is processed](https://github.com/apache/airflow/pull/10956/files/f3c6cf1dc10e509da1b3b1db2a595a58abfab847#diff-c35269bcfbbe386e269ffa7487e86192L723)), we can still create an extra DAG run here.

Reply:
The extra checks on max active runs etc have been added now.